File: 34-find-first-and-last-position-of-element-in-sorted-array/find-first-and-last-position-of-element-in-sorted-array.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        int lb = lowerBound(nums, 0, nums.length - 1, target);
        if(lb == nums.length || nums[lb] != target ) return new int []{-1, -1};
        return new int []{lb, upperBound(nums, 0, nums.length - 1, target) - 1};
    }
    public int lowerBound (int [] num, int low, int high , int x){
        int ans = num.length;
        while(low <= high){
            int mid= low + (high - low)/2;
            if(num[mid] >= x){
                ans = mid;
                high = mid - 1;
            }
            else{
                low = mid + 1;
            }
        }
        return ans;
    }
    public int upperBound (int [] num, int low, int high , int x){
        int ans = num.length;
        while(low <= high){
            int mid= low + (high - low)/2;
            if(num[mid] > x){
                ans = mid;
                high = mid - 1;
            }
            else{
                low = mid + 1;
            }
        }
        return ans;
    }
}

File: 34-find-first-and-last-position-of-element-in-sorted-array/find-first-and-last-position-of-element-in-sorted-array.cpp
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        int first = lower(nums, 0, nums.size() - 1, target);
        if(first == -1) return {-1, -1};
        int last = upper(nums, 0, nums.size() - 1, target);
        return {first, last};
    }
    int lower (vector <int> & nums, int low , int high , int x){
        int first = -1;
        while(low <= high){
            int mid = low + (high - low)/2;
            if(nums[mid] == x){
                first = mid;
                high = mid - 1;
            }
            else if(nums[mid] < x){
                low = mid + 1;
            }
            else{
                high = mid - 1;
            }
        }
        return first;
    }
    int upper (vector <int> & nums, int low , int high , int x){
        int last = -1;
        while(low <= high){
            int mid = low + (high - low)/2;
            if(nums[mid] == x){
                last = mid;
                low = mid + 1;
            }
            else if(nums[mid] < x){
                low = mid + 1;
            }
            else{
                high = mid - 1;
            }
        }
        return last;
    }
       
};